### PR TITLE
New metric module to improve flexibility and  intuitiveness

### PR DESCRIPTION
**Pull request by sanchit97**
# What does this PR do?

## Summary
*This PR adds a new metric module to Textattack, which provides an easier way to add new metrics or use existing metrics on adv. examples. Currently metrics are calculated in the `attack_log_manager.py` file, which is very unintuitive and non-flexible. Slides detailing this change are at: https://docs.google.com/presentation/d/1O77447GGWQF-xDnIdTr8d2MMPsQ9o9CG3pZZoo2bPrs/edit?usp=sharing*

## Additions
- *Added `metrics` module as `textattack.attack_metrics`.*

## Changes
- *`attack_metrics` is a new module, removing metrics from `attack_log_manager.py`*

## Deletions
- **

## Checklist
- [ x ] The title of your pull request should be a summary of its contribution.
- [ x ] Please write detailed description of what parts have been newly added and what parts have been modified. Please also explain why certain changes were made.
- [ x ] If your pull request addresses an issue, please mention the issue number in the pull request description to make sure they are linked (and people consulting the issue know you   are working on it)
- [ x ] To indicate a work in progress please mark it as a draft on Github.
- [ x ] Make sure existing tests pass.
- [ x ] Add relevant tests. No quality testing = no merge.
- [  ] All public methods must have informative docstrings that work nicely with sphinx. For new modules/files, please add/modify the appropriate `.rst` file in `TextAttack/docs/apidoc`.'

### CODE REVIEW

File: textattack/attack_metrics/attack_metric.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+========================
+
+"""
+
+from abc import ABC, abstractmethod
+
+from textattack.attack_results import AttackResult
+
+
+class AttackMetric(AttackResult, ABC):
```
Comment (jinyongyoo) [inline textattack/attack_metrics/attack_metric.py:12]: What's the reason why `AttackMetric` is a sub-class of `AttackResult`?

**Review by jinyongyoo — CHANGES_REQUESTED** [review]
Awesome start! Thank you for making this PR! I think it's good step towards cleaning up the messy calculations that happen within the log manager. Aside from the minor comments that I have made, the major question I have is about the way we/users interact with`AttackMetric`. 

Currently, for each list of attack results, we need to instantiate a fresh `AttackMetric` object. Within each `__init__` we call `calculate` that calculates the desirable metrics and then stores them as instance attributes (e.g. `self.total_attacks`). Then, we call various custom instance methods (e.g. `successful_attacks_num`, `avg_num_queries_num`) to retrieve the attributes.

I see two potential issues with this approach:

_(1) The need to instantiate a different `AttackMetric` object for each different list of results._
Why do need to create a new `AttackMetric` object for any new list of attack results? Is `AttackMetric`'s role **to store** the actual metric values for any arbitrary list of attack results, or is it to **calculate the metric** for a given list of results. If it is the latter, then I think there is no need to create a new instance. Instead, we can create one instance and then reuse it by passing a list of results to a method. This is similar to how our `GoalFunction` can be instantiated once and its `get_results` method can be used to produce `GoalFunctionResults` for list of `AttackedText`. 

_(2) Lack of common patterns between different `AttackMetric` classes._ 
We currently have three sub-classes of `AttackMetric` that share no common pattern. If I want to calculate the metric, I have to figure out which specific method I have to call (e.g.  `successful_attacks_num`, `avg_num_queries_num`) to get the actual values. This will inevitably become more confusing as we add more metric classes.

Potential Solution (this is only a suggestion, so please feel free to come up with your own solution): Have `calculate` method of `AttackMetric` take in a list of `AttackResult` as an argument and make it the main method that people can use to calculate their metrics. If there are multiple metrics we want to return, we can simply return a dictionary of the values with appropriate keys.

File: textattack/attack_metrics/attack_metric.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    """A metric for evaluating Adversarial Attack candidates."""
+
+    @staticmethod
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (qiyanjun) [reply textattack/attack_metrics/attack_metric.py:16]: @sanchit97 please update regarding the comment..

File: textattack/attack_metrics/attack_metric.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    def __init__(self, results, **kwargs):
+        """Creates pre-built :class:`~textattack.AttackMetric` that correspond to
+        evaluation metrics for adversarial examples.
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    @abstractmethod
+    def calculate():
```
[the remark this answers is not in the document]
Comment (qiyanjun) [reply textattack/attack_metrics/attack_metric.py:25]: @sanchit97 please update regarding the comment..

File: textattack/attack_metrics/attack_queries.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+from textattack.attack_results import SkippedAttackResult
+
+from .attack_metric import AttackMetric
+
+
+class AttackQueries(AttackMetric):
+    """Calculates all metrics related to number of queries in an attack
+
+    Args:
+    results (:obj::`list`:class:`~textattack.goal_function_results.GoalFunctionResult`):
```
[the remark this answers is not in the document]
Comment (qiyanjun) [reply textattack/attack_metrics/attack_queries.py:12]: @sanchit97 please update regarding the comment..

File: textattack/attack_metrics/attack_success_rate.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
+from textattack.attack_results import FailedAttackResult, SkippedAttackResult
+
+from .attack_metric import AttackMetric
+
+
+class AttackSuccessRate(AttackMetric):
+    """Calculates all metrics related to number of succesful, failed and skipped results in an attack
+
+    Args:
+    results (:obj::`list`:class:`~textattack.goal_function_results.GoalFunctionResult`):
```
[the remark this answers is not in the document]
Comment (qiyanjun) [reply textattack/attack_metrics/attack_success_rate.py:10]: @sanchit97 please update regarding the comment..

File: textattack/attack_metrics/attack_metric.py (hunk 1 above)
Comment (qiyanjun) [reply textattack/attack_metrics/attack_metric.py:12]: @sanchit97 I have the same concerns as @jinyongyoo above

**Comment by sanchit97** [comment]
My major objective to make different `AttackMetric` sub-classes was to group certain metrics together - e.g. `words_perturbed` only has metrics centered around word perturbations (like % of perturbed, max words changed, etc.). 
I thought if we add a new metric, let's say `Quality` we can add a new class that deals with the fluency of adversaries (perplexity), grammar errors, etc. It's a general grouping of the type of metrics.
However, your point is also true to simplify the process and add new functions for various metrics as we go and not group metrics together.

File: textattack/attack_metrics/attack_metric.py (hunk 1 above)
Comment (sanchit97) [reply textattack/attack_metrics/attack_metric.py:12]: Yes, this is an oversight. Fixed.

**Pull request closed without merging**